text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,289 @@
+
+None beyond the technical implications listed for `core`.
+
+## 4 - Allow the user to provide their own custom source versions of `core` and `std`
+
+### Use Case
+
+This will allow users of a nightly compiler to provide a custom version of `core` and `std`, without requiring the recompilation of the compiler itself.
+
+### Caveats
+
+As stability guarantees cannot be made around modified versions of `core` or `std`, a nightly compiler must always be used.
+
+### User Interaction
+
+For this interaction, the existing `patch` syntax of Cargo.toml will be used. For example:
+
+```toml
+[patch.sysroot]
+core = { path = 'my/local/core' }
+std  = { git = 'https://github.com/example/std' }
+```
+
+### Technical Implications
+
+The `patch.sysroot` term will be introduced for patch when referring to components such as `std` and `core`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This RFC introduces new concepts to the use of Rust and Cargo, and could be confusing for current users of Rust who have not had to consider changes to `core` or `std` previously. However, in the normal case, most users are unlikely to need these settings, while they allow users that DO need to make changes to control important steps of the build process.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+> Why is this design the best in the space of possible designs?
+
+This approach borrows from existing behaviors used by Cargo to allow configuration of `core` and `std`, as if they were a regular crate dependency.
+
+This approach also offers an approach that can be developed and applied incrementally, allowing for time to find coner cases not considered by this RFC
+
+> What other designs have been considered and what is the rationale for not choosing them?
+
+To the author of this RFC's knowledge, there are no other open designs, other than the use tools that wrap Cargo entirely, such as [xargo].
+
+[xargo]: https://github.com/japaric/xargo
+
+> What is the impact of not doing this?
+
+By not doing this, Rust will continue to be difficult to use for users and platforms "on the edge", such as new platform developers or embedded and WASM users.
+
+# Prior art
+[prior-art]: #prior-art
+
+* https://github.com/rust-lang/rfcs/pull/1133
+* https://github.com/japaric/xargo
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## How are dependencies for `core` and `std` specified?
+
+For example in a `no_core` or `no_std` crate, how would we tell Cargo **not** to build the `core` and/or `std` dependencies?
+
+## Should `std` be rebuilt if `core` is rebuilt?
+
+Is it necessary to rebuild `std` using the customized `core`, even if no changes to `std` are necessary?
+
+## Should Cargo obtain or verify the source code for `libcore` or `libstd`?
+
+Right now we depend on `rustup` to obtain the correct source code for these libraries, and we rely on the user not to tamper with the contents. Are these reasonable decisions?
```

Comment:
Discuss possible ways to avoid tampering? If we rely on users not tampering, how do we communicate this effectively?

Reply:
I guess the first thing we should decide is IF we should try to detect tampering, and whether that makes sense when the user owns their own PC anyway.

I am of the opinion that we shouldn't (as I think it is a cat and mouse game), however this was brought up by multiple people :)

Reply:
> I guess the first thing we should decide is IF we should try to detect tampering, and whether that makes sense when the user owns their own PC anyway.

Right, but we should at least be sure that we can detect tampering.. ;) IOW, let's not decide we want it and find out later that we cannot.

> I am of the opinion that we shouldn't (as I think it is a cat and mouse game), however this was brought up by multiple people :)

Things like people wanting to use `RUSTC_BOOTSTRAP` on stable make me less sure; but maybe we can use social instead of technical means to discourage tampering?

Reply:
IMO: Changing libcore/libstd would only affect the binaries produced by building this project, and would not be easy to "accidentally" end up with (except as a consumer of that binary). At the end of the day, anyone can patch the rust compiler for their own builds and distributions, which would defeat any measures we put in place.

Even if we bake in the CRCs of the source files into `rustc`, people can patch the `rustc` binary, or rebuild their own compiler. At the end of the day, the compiler is code running on *their* computer, which we can't do much about.

At best, I think tamper detection would serve only as a warning, "you changed the source, this is not supported behavior".

Reply:
> At best, I think tamper detection would serve only as a warning, "you changed the source, this is not supported behavior".

If it's not too troublesome performance and implementation-wise, that seems like a decent solution; at least we have communicated what we do and do not consider stable in a direct way then. :)

Reply:
Sure! I am just trying to avoid wasted engineering time, and even just baking the CRC of the total source into the compiler seems like it may be more trouble than it is worth. Let's see what other people's feedback is!

Reply:
Yeah fair; for now you could discuss this as a *possibility* :)